### PR TITLE
API: avoid passing Manager to subclass __init__

### LINKED ISSUE #57032: BUG: DeprecationWarning shown for internal operations on subclassed Series and DataFrame
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import warnings

import pandas as pd

warnings.simplefilter('always', DeprecationWarning)


class MySeries(pd.Series):

    @property
    def _constructor(self):
        return MySeries


series = MySeries({'A': 1, 'B': 2, 'C': 3})

series.loc[['A', 'B']]
```

### Issue Description

The above code will print 
```
DeprecationWarning: Passing a SingleBlockManager to MySeries is deprecated and will raise in a future version. Use public APIs instead.
```
because ``Series._constructor_from_mgr`` calls ``self._constructor(mgr)`` if ``self._constructor`` is not a Series without catching that warning. 
https://github.com/pandas-dev/pandas/blob/72814750548b5ead5b08cd9b90d56a23c9a520ea/pandas/core/series.py#L665-L673

The warning is not very helpful in this case, because it recommends to use public APIs, but only public APIs are used. There's also no way to fix it except for just filtering it out.

### Expected Behavior

Pandas should not raise this warning when it is actually caused by an internal operation that cannot be changed by a user.

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit                : 72814750548b5ead5b08cd9b90d56a23c9a520ea
python                : 3.12.1.final.0
python-bits           : 64
OS                    : Windows
OS-release            : 10
Version               : 10.0.19044
machine               : AMD64
processor             : AMD64 Family 23 Model 113 Stepping 0, AuthenticAMD
byteorder             : little
LC_ALL                : None
LANG                  : None
LOCALE                : English_United Kingdom.1252
pandas                : 3.0.0.dev0+163.g7281475054
numpy                 : 1.26.3
pytz                  : 2023.3.post1
dateutil              : 2.8.2
setuptools            : 69.0.3
pip                   : 23.3.2
Cython                : None
pytest                : 7.4.4
hypothesis            : None
sphinx                : 7.2.6
blosc                 : None
feather               : None
xlsxwriter            : None
lxml.etree            : None
html5lib              : None
pymysql               : None
psycopg2              : None
jinja2                : 3.1.2
IPython               : None
pandas_datareader     : None
adbc-driver-postgresql: None
adbc-driver-sqlite    : None
bs4                   : 4.12.2
bottleneck            : None
dataframe-api-compat  : None
fastparquet           : None
fsspec                : None
gcsfs                 : None
matplotlib            : 3.8.2
numba                 : None
numexpr               : None
odfpy                 : None
openpyxl              : None
pandas_gbq            : None
pyarrow               : None
pyreadstat            : None
python-calamine       : None
pyxlsb                : None
s3fs                  : None
scipy                 : 1.12.0
sqlalchemy            : None
tables                : None
tabulate              : None
xarray                : None
xlrd                  : None
zstandard             : None
tzdata                : 2023.4
qtpy                  : None
pyqt5                 : None

</details>

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #57032 (Replace xxxx with the GitHub issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

Many subclasses have `__init__` methods that call `super().__init__`, which in the relevant cases with a Manager object is deprecated.  The solution in this PR is to construct the pd.DataFrame/Series and pass that to the subclass's `_constructor`.

This removes _sliced_from_mgr and _expanddim_from_mgr, as they are not really necessary, simplifying things a bit.  Also makes _from_mgr `@final`.  Can separate these simplifications into their own PR on request.  xref #56681.

For subclasses that currently have special handling for Manager objects in their `__init__`, they will need to move that special handling to the appropriate `_constructor(_foo)?_from_mgr` method.  The only package I'm aware of that does this is geopandas, so this includes a shim special-casing GeoDataFrame intended to maintain the current behavior until they have time to move that logic.  cc @jorisvandenbossche ATM the shim is only in _constructor_from_mgr, pls advise on whether something analogous is needed in constructor_sliced_from_mgr and _constructor_expanddim_from_mgr.

### CODE REVIEW

File: pandas/core/frame.py [hunk 1]
```diff
@@ -650,25 +650,36 @@ def _constructor(self) -> type[DataFrame]:
...
-        else:
-            assert axes is mgr.axes
+        df = DataFrame._from_mgr(mgr, axes=axes)
+
+        if type(self) is DataFrame:
+            # This would also work `if self._constructor is DataFrame`, but
+            #  this check is slightly faster, benefiting the most-common case.
+            return df
+
+        elif type(self).__name__ == "GeoDataFrame":
```
Comment (mroeschke) [inline pandas/core/frame.py:660]: Worth having a test for this in `test_downstream.py`? `geopandas` is installed in the CI

Comment (jbrockmendel) [reply pandas/core/frame.py:660]: Good idea (also should add a test for#57032).  Will wait until Joris weighs in

Comment (jbrockmendel) [reply pandas/core/frame.py:660]: im having trouble writing a meaningful test for this cc @jorisvandenbossche

Comment (jorisvandenbossche) [reply pandas/core/frame.py:661]: GeoPandas is already overriding `_constructor_from_mgr`, so normally this shouldn't be needed.

(will try to take a closer look one of the coming days!)

**Review by phofl — APPROVED** [review]

**Comment by phofl** [comment]
thx @jbrockmendel

**Pull request merged**